_posts/2019-07-04-exploring-nservicebus-integration-testing-options.md

```diff
@@ -0,0 +1,167 @@
+            )
+            || c.HasFailedMessages();
+        })
+        .Run();
+
+    var invokedSaga = context.InvokedSagas.Single(s => s.SagaType == typeof(ASaga));
+
+    Assert.True(invokedSaga.IsNew);
+    Assert.True(((ASagaData)invokedSaga.SagaData).SomeId == expectedSomeId);
+    Assert.False(context.HasFailedMessages());
+    Assert.False(context.HasHandlingErrors());
+}
+```
+
+There is lot going on, I'd like to highlight just two things for now:
+
+- The test is spinning up 2 endpoints, `MyServiceEndpoint` and `MyOtherServiceEndpoint`. Those are real NServiceBus endpoints.
+- The test is defining a `Done` condition used internally by the testing infrastructure to determine that the test is done and asserts can be executed. Thus is needed because there real messages flowing around on real queues, so everything is asynchronous. What's interesting is what's in the done condition, the test is done:
+  - `AMessageHandler` is invoked (satisfying `MyService` sends `AMessage` to `MyOtherService`)
+  - `AReplyMessageHandler` is invoked (satisfying`MyService` handles `AReplyMessage`)
+  - `ASaga` is invoked (satisfying `ASaga` is started in `MyService`)
+  - Or something goes badly and there are errors
+
+### Endpoints
+
+What are those endpoints? They are defined in the same class as the test, as follows:
+
+```csharp
+class MyServiceEndpoint : EndpointConfigurationBuilder
+{
+   public MyServiceEndpoint()
+   {
+      EndpointSetup<ServiceTemplate<MyServiceConfiguration>>();
+   }
+}
+```
+
+I'm gaming the system a little bit here. Under the hood I'm still using the NServiceBus Acceptance Testing framework for now as it has all the infrastructure required to run endpoints, but I need to be able to run endpoints whose configuration is defined elsewhere. It has to be the user endpoint configuration. `MyServiceConfiguration` is defined in the `MyService` endpoint project and is the configuration that will end up being used in production. It looks like this:
+
+```csharp
+public class MyServiceConfiguration : EndpointConfiguration
+{
+    public MyServiceConfiguration()
+        : base("MyService")
+    {
+        this.UseSerialization<NewtonsoftSerializer>();
+        this.UsePersistence<LearningPersistence>();
+        var transportConfig = this.UseTransport<LearningTransport>();
+
+        transportConfig.Routing()
+            .RouteToEndpoint(typeof(AMessage), "MyOtherService");
+
+        this.SendFailedMessagesTo("error");
+
+        //omitted for clarity
+    }
+}
+```
+
+Here is the trick: in the endpoint code instead of using the `EndpointConfiguration` class directly I'm creating a class the inherits from it setting configuration values in the constructor. In this way I can then reference that class in the test to identify a configuration that should be used to spin up an endpoint. The `ServiceTemplate` thing in the test code is again a tricky way to glue my spike with the existing Acceptance testing infrastructure.
```

Comment:
```suggestion
Here is the trick: in the endpoint code instead of using the `EndpointConfiguration` class directly I'm creating a class the inherits from it setting configuration values in the constructor. This way I can then reference that class in the test to identify a configuration that should be used to spin up an endpoint. The `ServiceTemplate` thing in the test code is again a tricky way to glue my spike with the existing Acceptance testing infrastructure.
```